### ISSUE TITLE
Installing bumblebee-nvidia will break intel driver configuration.

### ISSUE THREAD

**Issue by akiross** [gsgatlin/optimus-rpms#1]
I hope this is the right place to post an issue. I'm following [this guide](https://fedoraproject.org/wiki/Bumblebee#For_closed_source_solution_fedora). I have been using bumblebee-nvidia successfully for a while, but after a recent (few days ago) update something is messed-up. I can't understand where the problem is.

I am on Fedora 25, trying to install managed bumblebee-nvidia 375.39-2.

Starting from a system without bumblebee installed, only Intel drivers, I can run `glxinfo` without any problem, and I can retrieve the right OpenGL drivers:

```
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) HD Graphics 520 (Skylake GT2) 
OpenGL core profile version string: 4.5 (Core Profile) Mesa 13.0.4
```

Everything is working correctly with the Intel card.

If I install managed bumblebee-nvidia, the outcome may vary. I am serious here: starting from what I believe is a valid intel-only system configuration, after bumblebee-nvidia is installed, the outcome is random. Until now, I experienced the following:

1. after installing bumblebee-nvidia, I restart the computer and the driver is compiled automatically. After it finishes, `bumblebee-nvidia -c` says module is compiled Ok, and in fact I can run `optirun glxinfo` and get NVIDIA OpenGL reported. The problem is that **glxinfo won't work without optirun!**. The old intel driver has been corrupted in some way:
```
$ glxinfo
name of display: :0
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  151 (GLX)
  Minor opcode of failed request:  24 (X_GLXCreateNewContext)
  Value in failed request:  0x0
  Serial number of failed request:  35
  Current serial number in output stream:  36
```

2. system reboots and hangs, because wayland won't show up. In this case, I only proceeded by entering rescue mode and removing bumblebee\*, and this apparently fixes everything. Sometimes, I had to reinstall 32bit version of libGL because it was broken.

I need 32+64bit libraries, but I experience this problems even using only 64bit version.

I don't know what to do.

**EDIT** I would like to add that, when glxinfo is broken, reinstalling mesa-libGL and intel drivers will not fix the issue. I thought it could be a problem of wrong library referenced, and I run the following:

```
$ ldd `which glxinfo`
        linux-vdso.so.1 (0x00007ffc351a6000)
        libGL.so.1 => /usr/lib64/nvidia-bumblebee/libGL.so.1 (0x00007fe51b0e2000)
        libX11.so.6 => /lib64/libX11.so.6 (0x00007fe51ad75000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fe51a9af000)
        libnvidia-tls.so.375.39 => /usr/lib64/nvidia-bumblebee/tls/libnvidia-tls.so.375.39 (0x00007fe51a7ab000)
        libnvidia-glcore.so.375.39 => /usr/lib64/nvidia-bumblebee/libnvidia-glcore.so.375.39 (0x00007fe5189d6000)
        libXext.so.6 => /lib64/libXext.so.6 (0x00007fe5187c4000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007fe5185c0000)
        libxcb.so.1 => /lib64/libxcb.so.1 (0x00007fe518398000)
        /lib64/ld-linux-x86-64.so.2 (0x00005616e983e000)
        libm.so.6 => /lib64/libm.so.6 (0x00007fe51808f000)
        libXau.so.6 => /lib64/libXau.so.6 (0x00007fe517e8b000)
```
Is everything right, here? Are nvidia-bumblebee/lib*.so expected to be used even without optirun?

**Comment by gsgatlin** [comment]
Hello. Recently there were some changes in fedora. and Also the driver had to be changed due to a kernel upgrade. But perhaps we can figure out what is wrong in the logs. I am using f25 but I am using MATE spin instead of wayland with GNOME.

I have a test system here though with an older card which has wayland + GNOME.


I just upgraded it this morning and everything appears to work. Here is the output from both (working) machines. (Lenovo Thinkpad t540p using MATE and Lenovo Ideapad y470 using GNOME + wayland)

https://paste.fedoraproject.org/paste/OfxWVSow9iIxGI7cDEuQil5M1UNdIGYhyRLivL9gydE=

Can you post to http://pastebin.com/ or https://paste.fedoraproject.org/ or similar site your

/var/log/nvidia-installer.log

And post the link to your log file? That is, is after it appears to work but messes up your intel libs.

**Comment by akiross** [comment]
Hello, I suspected changes in distro and drivers, but my current knowledge does not allow me to search and understand where the issue could be.

https://paste.fedoraproject.org/paste/CQVsDmO3Dpqjz6PCNs45lV5M1UNdIGYhyRLivL9gydE=

I am currently using the system with bumblebee-nvidia installed.

Also, I am a bit puzzled because I thought that GNOME Shell on Wayland used OpenGL, but since glxinfo and glxgears do not work, I expected OGL application not to run at all. Also, nvidia module was *not* loaded after the boot, so I don't understand why things work (even steam).

**Comment by gsgatlin** [comment]
hmnn. Do you have VirtualGL.i686 package installed?

[gsgatlin@y470c ~]$ rpm -q VirtualGL.i686
VirtualGL-2.4-7.fc25.i686


What is the output from?

ls -l /usr/lib/lib*GL*

ls -l /usr/lib64/lib*GL*

Do you know if you are using X? maybe it falls back to that?

do you have a /var/log/Xorg.0.log ?

**Comment by akiross** [comment]
Yes, I have VirtualGL installed 
```
$ rpm -q VirtualGL
VirtualGL-2.4-7.fc25.i686
VirtualGL-2.4-7.fc25.x86_64
$ sudo dnf install VirtualGL
Last metadata expiration check: 0:03:44 ago on Wed Mar  1 17:05:41 2017.
Package VirtualGL-2.4-7.fc25.i686 is already installed, skipping.
Package VirtualGL-2.4-7.fc25.x86_64 is already installed, skipping.
Dependencies resolved.
Nothing to do.
Complete!
```

I think I am using Wayland:
```
$ echo $WAYLAND_DISPLAY 
wayland-0
$ loginctl show-session 2 -p Type
Type=wayland
```

I do have a [/var/log/Xorg.0.log](https://paste.fedoraproject.org/paste/K9MCZDMpuqByE95gjfl8V15M1UNdIGYhyRLivL9gydE=).

Right now I am experimenting with short lived driver branch and unmanaged repo to see if something changes, so I cannot tell you which libGL are being used, because I already uninstalled the nvidia driver.
I will report shortly if there were any changes.

**Comment by akiross** [comment]
So, tried with 378.13 and basically nothing changed, but at least I had the idea to check better the system state during the installation process.

Here's the step I did:

1. completely uninstalled bumblebee and its repos with `sudo dnf remove bumblebee\*`
2. then I added the repos **bumblebee-release-1.2-1** and **bumblebee-nonfree-unmanaged-release-1.2-1**
3. after downloading the nvidia drivers blob, I copied it to /etc/sysconfig/nvidia/ `sudo cp Downloads/NVIDIA-Linux-x86_64-378.13.run /etc/sysconfig/nvidia/`
4. reinstalled intel drivers and libs: `sudo dnf reinstall libglvnd-glx\* xorg-x11-drv-intel\* mesa-libGL\*`
5. rebooted and made sure that `nvidia.ko` was not present and everything worked. glxinfo worked and pointed to mesa 13, also
```
$ ldd `which glxinfo`
        linux-vdso.so.1 (0x00007fffd04f9000)
        libGL.so.1 => /lib64/libGL.so.1 (0x00007f5dfa0e7000)
        libX11.so.6 => /lib64/libX11.so.6 (0x00007f5df9da8000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f5df99e2000)
        libGLX.so.0 => /lib64/libGLX.so.0 (0x00007f5df97b1000)
        libXext.so.6 => /lib64/libXext.so.6 (0x00007f5df959d000)
        libGLdispatch.so.0 => /lib64/libGLdispatch.so.0 (0x00007f5df92e7000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f5df90e3000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f5df8ec5000)
        libxcb.so.1 => /lib64/libxcb.so.1 (0x00007f5df8c9d000)
        /lib64/ld-linux-x86-64.so.2 (0x00005630ee53b000)
        libXau.so.6 => /lib64/libXau.so.6 (0x00007f5df8a99000)
```

6. reinstalled bumblebee-nvidia (with compatibility)
```
sudo dnf install bumblebee-nvidia bbswitch-dkms VirtualGL.x86_64 VirtualGL.i686 primus.x86_64 primus.i686 kernel-devel
```
7. rebooted
8. nvidia kernel module started to compile, and *before* it was compiled `glxinfo` worked, pointing to mesa 13 as before.
9. after nvidia module was compiled, `glxinfo` was not running anymore, but `optirun glxinfo` do work. Again:
```
$ ldd `which glxinfo`
        linux-vdso.so.1 (0x00007fffc9d37000)
        libGL.so.1 => /usr/lib64/nvidia-bumblebee/libGL.so.1 (0x00007f8bb05e1000)
        libX11.so.6 => /lib64/libX11.so.6 (0x00007f8bb0274000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f8bafeae000)
        libnvidia-tls.so.378.13 => /usr/lib64/nvidia-bumblebee/tls/libnvidia-tls.so.378.13 (0x00007f8bafcaa000)
        libnvidia-glcore.so.378.13 => /usr/lib64/nvidia-bumblebee/libnvidia-glcore.so.378.13 (0x00007f8badeb2000)
        libXext.so.6 => /lib64/libXext.so.6 (0x00007f8badca0000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f8bada9c000)
        libxcb.so.1 => /lib64/libxcb.so.1 (0x00007f8bad874000)
        /lib64/ld-linux-x86-64.so.2 (0x0000559069663000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f8bad56b000)
        libXau.so.6 => /lib64/libXau.so.6 (0x00007f8bad367000)
```

So, apparently something in the installer is messing up. Here's my new [/var/log/nvidia-installer.log](https://paste.fedoraproject.org/paste/nRkK~X56M9eK3~sff~FjM15M1UNdIGYhyRLivL9gydE=).

**EDIT:** forgot to add libraries:

```
$ ll /usr/lib/libGL.*   
lrwxrwxrwx. 1 root root   10 Jan  9 20:39 /usr/lib/libGL.so -> libGL.so.1
lrwxrwxrwx. 1 root root   14 Feb  6 14:09 /usr/lib/libGL.so.1 -> libGL.so.1.0.0
-rwxr-xr-x. 1 root root 400K Feb  6 14:09 /usr/lib/libGL.so.1.0.0
$ ll /usr/lib64/libGL.*
lrwxrwxrwx. 1 root root   14 Feb  6 14:09 /usr/lib64/libGL.so -> libGL.so.1.0.0
lrwxrwxrwx. 1 root root   14 Feb  6 14:09 /usr/lib64/libGL.so.1 -> libGL.so.1.0.0
-rwxr-xr-x. 1 root root 569K Feb  6 14:09 /usr/lib64/libGL.so.1.0.0
```

**Comment by gsgatlin** [comment]
The only thing I noticed in teh log that was different was the line:

executing: '/usr/sbin/ldconfig'...
   /usr/sbin/ldconfig: libraries libGL.so.1 and libGL.so.1.0.0 in directory /lib have same soname but different type.
   executing: '/usr/sbin/depmod -a '...
-> done.

which is in mine but not yours.

Here is an avenue we can try.

[gsgatlin@y470c ~]$ ls -ltr /usr/lib64 

The end of the listing looks like this:

...
-rwxr-xr-x.  1 root root    16096 Feb 25 19:21 libfsimage.so.1.0.0
-rwxr-xr-x.  1 root root    32528 Feb 25 19:21 libblktapctl.so.1.0.0
-rwxr-xr-x.  1 root root    54736 Feb 27 04:26 libpci.so.3.5.4
lrwxrwxrwx.  1 root root       15 Feb 27 04:26 libpci.so.3 -> libpci.so.3.5.4
drwxr-xr-x.  4 root root     4096 Mar  1 09:55 security
drwxr-xr-x.  4 root root     4096 Mar  1 09:56 tracker-1.0
drwxr-xr-x.  2 root root     4096 Mar  1 09:59 grilo-0.3
drwxr-xr-x.  2 root root     4096 Mar  1 09:59 fipscheck
drwxr-xr-x.  2 root root     4096 Mar  1 09:59 rpm-plugins
drwxr-xr-x.  2 root root    12288 Mar  1 09:59 gstreamer-1.0
drwxr-xr-x.  2 root root    12288 Mar  1 10:00 girepository-1.0
-rwxr-xr-x.  1 root root  1805096 Mar  1 10:07 libnvidia-gtk2.so.375.39
-rwxr-xr-x.  1 root root  1799336 Mar  1 10:07 libnvidia-gtk3.so.375.39
-rwxr-xr-x.  1 root root   157896 Mar  1 10:07 libnvidia-cfg.so.375.39
-rwxr-xr-x.  1 root root    94440 Mar  1 10:07 libnvidia-fbc.so.375.39
lrwxrwxrwx.  1 root root       23 Mar  1 10:07 libnvidia-cfg.so.1 -> libnvidia-cfg.so.375.39
drwxr-xr-x.  5 root root     4096 Mar  1 10:07 nvidia-bumblebee
lrwxrwxrwx.  1 root root       23 Mar  1 10:07 libnvidia-fbc.so.1 -> libnvidia-fbc.so.375.39
lrwxrwxrwx.  1 root root       18 Mar  1 10:07 libnvidia-fbc.so -> libnvidia-fbc.so.1
lrwxrwxrwx.  1 root root       18 Mar  1 10:07 libnvidia-cfg.so -> libnvidia-cfg.so.1
drwxr-xr-x.  2 root root     4096 Mar  1 10:07 vdpau
drwxr-xr-x.  2 root root     4096 Mar  1 10:07 fakelib

ls -ltr /usr/lib

...
-rwxr-xr-x.  1 root root    73524 Feb 22 06:53 libnss_myhostname.so.2
lrwxrwxrwx.  1 root root       16 Feb 23 03:05 libpcre.so.1 -> libpcre.so.1.2.8
lrwxrwxrwx.  1 root root       21 Feb 23 03:05 libpcreposix.so.0 -> libpcreposix.so.0.0.4
-rwxr-xr-x.  1 root root   483404 Feb 23 03:05 libpcre.so.1.2.8
-rwxr-xr-x.  1 root root    10352 Feb 23 03:05 libpcreposix.so.0.0.4
drwxr-xr-x.  2 root root     4096 Mar  1 09:55 sysctl.d
drwxr-xr-x.  2 root root     4096 Mar  1 09:55 sysusers.d
drwxr-xr-x.  2 root root     4096 Mar  1 09:55 ostree
drwxr-xr-x.  4 root root     4096 Mar  1 09:56 udev
drwxr-xr-x. 14 root root     4096 Mar  1 09:59 systemd
drwxr-xr-x.  2 root root     4096 Mar  1 09:59 tmpfiles.d
drwxr-xr-x.  5 root root     4096 Mar  1 10:01 modules
drwxr-xr-x.  5 root root     4096 Mar  1 10:01 rpm
drwxr-xr-x.  4 root root     4096 Mar  1 10:07 nvidia-bumblebee
drwxr-xr-x.  2 root root     4096 Mar  1 10:07 fakelib

Do you have different files that were modified today?

**Comment by akiross** [comment]
The only relevant different thing seems to be the version.
 
```
$ ls -ltr /usr/lib | tail -n 15      
drwxr-xr-x.  2 root root     4096 Feb 27 10:43 ostree
drwxr-xr-x. 14 root root     4096 Feb 27 10:44 systemd
drwxr-xr-x.  2 root root     4096 Feb 27 10:45 girepository-1.0
drwxr-xr-x.  2 root root     4096 Feb 27 10:45 gstreamer-1.0
drwxr-xr-x.  5 root root     4096 Feb 27 12:12 rpm
drwxr-xr-x.  4 root root     4096 Feb 27 23:36 udev
drwxr-xr-x.  2 root root     4096 Feb 27 23:36 kdump
drwxr-xr-x.  9 root root     4096 Feb 28 16:14 modules
drwxr-xr-x.  2 root root     4096 Feb 28 16:14 tmpfiles.d
drwxr-xr-x.  2 root root     4096 Feb 28 19:06 gallium-pipe
drwxr-xr-x.  2 root root     4096 Feb 28 19:06 dri
drwxr-xr-x.  2 root root     4096 Mar  1 17:23 primus
drwxr-xr-x.  4 root root     4096 Mar  1 17:28 nvidia-bumblebee
drwxr-xr-x.  2 root root     4096 Mar  1 17:28 vdpau
drwxr-xr-x.  2 root root     4096 Mar  1 17:28 fakelib
$ ls -ltr /usr/lib64 | tail -n 15
drwxr-xr-x.  2 root root     4096 Feb 28 19:06 gallium-pipe
drwxr-xr-x.  2 root root     4096 Feb 28 19:06 dri
drwxr-xr-x.  2 root root    12288 Mar  1 17:08 pkgconfig
drwxr-xr-x.  2 root root     4096 Mar  1 17:23 primus
-rwxr-xr-x.  1 root root  1811400 Mar  1 17:28 libnvidia-gtk2.so.378.13
-rwxr-xr-x.  1 root root  1805160 Mar  1 17:28 libnvidia-gtk3.so.378.13
-rwxr-xr-x.  1 root root   158888 Mar  1 17:28 libnvidia-cfg.so.378.13
-rwxr-xr-x.  1 root root    94728 Mar  1 17:28 libnvidia-fbc.so.378.13
drwxr-xr-x.  5 root root     4096 Mar  1 17:28 nvidia-bumblebee
lrwxrwxrwx.  1 root root       23 Mar  1 17:28 libnvidia-fbc.so.1 -> libnvidia-fbc.so.378.13
lrwxrwxrwx.  1 root root       18 Mar  1 17:28 libnvidia-fbc.so -> libnvidia-fbc.so.1
lrwxrwxrwx.  1 root root       23 Mar  1 17:28 libnvidia-cfg.so.1 -> libnvidia-cfg.so.378.13
lrwxrwxrwx.  1 root root       18 Mar  1 17:28 libnvidia-cfg.so -> libnvidia-cfg.so.1
drwxr-xr-x.  2 root root     4096 Mar  1 17:28 vdpau
drwxr-xr-x.  2 root root     4096 Mar  1 17:28 fakelib
```

**Comment by akiross** [comment]
HA! Found the issue: LD_LIBRARY_PATH was pointing to `/usr/lib64/nvidia-bumblebee/`, this is why glxinfo was referring to it! Anyway, installation seems to work properly now, even if I don't quite understand why, as happened, wayland would not start at all... Possibly I left a corrupted state on the system while experimenting (maybe I did not correctly reinstall intel libs after removing bumblebee, or maybe it's because I messed with symlinks as suggested by the installer).

I'll close the issue, thanks for your help!
